File: spec/ui/CellMechanics.hs
-- CellMechanics.hs
-- Tesseract spec — UI layer
--
-- THE CONTROL-FACE ALGEBRA: every interactive region wears a face whose
-- state is an OPAQUE ink transform, never alpha. Ported from SixFour's
-- Spec.CellMechanics; the Swift twin is hand-checked
-- (Tesseract/UI/Cells/CellMechanics.swift) and golden-pinned by
-- TesseractTests/CellMechanicsParityTests.
--
--   states     = idle | pressed | busy | disabled          (ordinals 0-3)
--   treatments = ghost | lit | inverted | busy | checker   (ordinals 0-4)
--   faces      = frame | brackets
--
--   treatment(state, tick) = table[state·2 + beat]
--     where beat = 1 iff state == idle AND tick ≡ 0 (mod 4)
--
--   UM1  table shape        8 entries = |states| × 2
--   UM2  treatments total   every entry names a real treatment ordinal —
--                           the lawFaceNoAlpha carrier: each treatment is
--                           an opaque ink, so a total table means no
--                           state can ever fall through to alpha
--   UM3  only idle ticks    pressed/busy/disabled are tick-invariant
--   UM4  beat period        beat true exactly on tick ≡ 0 (mod 4) — the
--                           5 Hz BEAT at the 20 Hz clock; goldenBeat[1]
--                           is False, the reduce-motion anchor
--   UM5  idle semantics     idle is ghost off-beat, lit on-beat
--   UM6  golden table       the flat table equals SixFour's shipped
--                           [0,1,2,2,3,3,4,4] (the cross-app pin)
--
-- TRUTHFUL harness: every law asserted, exits nonzero on failure.

module Main where

import System.Exit (exitFailure, exitSuccess)

data ControlState = Idle | Pressed | Busy | Disabled
  deriving (Eq, Ord, Enum, Bounded, Show)

data Treatment = Ghost | Lit | Inverted | BusyInk | Checker
  deriving (Eq, Ord, Enum, Bounded, Show)

beatPeriod :: Int
beatPeriod = 4

-- | The semantic definition — treatment as a function of state and beat.
treatmentOf :: ControlState -> Bool -> Treatment
treatmentOf Idle     beat = if beat then Lit else Ghost
treatmentOf Pressed  _    = Inverted
treatmentOf Busy     _    = BusyInk
treatmentOf Disabled _    = Checker

-- | The flat table the Swift twin ships: index = state·2 + (beat?1:0).
table :: [Int]
table = [ fromEnum (treatmentOf s b)
        | s <- [minBound .. maxBound], b <- [False, True] ]

beat :: ControlState -> Int -> Bool
beat s t = s == Idle && t `mod` beatPeriod == 0

-- | The full evaluator, as the Swift `faceTreatment(state:tick:)`.
faceTreatment :: ControlState -> Int -> Int
faceTreatment s t = table !! (fromEnum s * 2 + (if beat s t then 1 else 0))

goldenBeat :: [Bool]
goldenBeat = [ t `mod` beatPeriod == 0 | t <- [0 .. 15] ]

-- ════════════════════════════════════════════════════════════════
-- Laws
-- ════════════════════════════════════════════════════════════════

um1_shape :: Bool
um1_shape = length table == length ([minBound .. maxBound] :: [ControlState]) * 2

um2_total :: Bool
um2_total = all (\e -> e >= 0 && e <= fromEnum (maxBound :: Treatment)) table

um3_onlyIdleTicks :: Bool
um3_onlyIdleTicks = and
  [ faceTreatment s t1 == faceTreatment s t2
  | s <- [Pressed, Busy, Disabled], t1 <- [0 .. 15], t2 <- [0 .. 15] ]

um4_beatPeriod :: Bool
um4_beatPeriod = goldenBeat == [ t `mod` 4 == 0 | t <- [0 .. 15] ]
              && not (goldenBeat !! 1)

um5_idle :: Bool
um5_idle = and [ faceTreatment Idle t
                 == fromEnum (if t `mod` 4 == 0 then Lit else Ghost)
               | t <- [0 .. 15] ]

um6_golden :: Bool
um6_golden = table == [0, 1, 2, 2, 3, 3, 4, 4]

check :: String -> Bool -> IO Bool
check name ok = do
  putStrLn ((if ok then "  ✓ " else "  ✗ ") ++ name)
  return ok

main :: IO ()
main = do
  putStrLn "CellMechanics: the control-face algebra (opaque ink, one clock)"
  results <- sequence
    [ check "UM1 table shape: 8 = |states| × 2"                  um1_shape
    , check "UM2 treatments total (lawFaceNoAlpha carrier)"       um2_total
    , check "UM3 only idle reads the tick"                        um3_onlyIdleTicks
    , check "UM4 beat ≡ tick mod 4 == 0; goldenBeat[1] = False"   um4_beatPeriod
    , check "UM5 idle = ghost off-beat, lit on-beat"              um5_idle
    , check "UM6 golden table [0,1,2,2,3,3,4,4]"                  um6_golden
    ]
  putStrLn ("  golden table  = " ++ show table)
  putStrLn ("  golden beat   = " ++ show (map fromEnum goldenBeat))
  if and results then exitSuccess else exitFailure

File: Tesseract/Surface/Cells/CellMechanics.swift
import Foundation
import simd

/// The closed control-face algebra (hand-checked twin of
/// `spec/ui/CellMechanics.hs`; golden-pinned by `CellMechanicsParityTests`).
/// Ported from SixFour's CellMechanicsContract.
///
/// Every interactive region wears a FACE (frame ring or corner brackets)
/// whose treatment is an OPAQUE ink transform — never alpha:
///   idle     → ghost, going LIT for exactly 1 tick in 4 (the BEAT);
///   pressed  → full ink inversion;
///   busy     → reject red;
///   disabled → the 2×2 checker.
/// Only idle reads the tick; all other states are tick-invariant.
enum CellMechanics {
    /// Control state ordinals.
    static let controlStates = ["idle", "pressed", "busy", "disabled"]
    /// Face treatment ordinals (opaque ink transforms).
    static let faceTreatments = ["ghost", "lit", "inverted", "busy", "checker"]
    /// Face geometries: "frame" = 1-cell inset ring (solid controls);
    /// "brackets" = corner brackets in the gutter (image-content controls).
    static let faceKinds = ["frame", "brackets"]

    /// The BEAT period: lit 1 tick in 4 (5 Hz at the 20 Hz clock).
    static let beatPeriodTicks = 4

    /// treatment = table[state × 2 + (beat ? 1 : 0)].
    /// Row layout: idle→(ghost|lit), pressed→(inverted|inverted),
    /// busy→(busy|busy), disabled→(checker|checker).
    static let faceTreatmentTable = [0, 1, 2, 2, 3, 3, 4, 4]

    /// The 16-tick golden beat pin: true exactly when tick ≡ 0 (mod 4).
    /// `goldenBeat[1] == false` is the reduce-motion anchor — evaluating
    /// any treatment at tick 1 is provably beat-free.
    static let goldenBeat: [Bool] = (0..<16).map { $0 % 4 == 0 }

    /// Face inks (re-exported from `Ink` so the algebra is self-contained).
    static let faceControlInk = Ink.ink
    static let rejectInk = Ink.reject
    static let acceptInk = Ink.accept

    /// The face treatment for a control state at a clock tick.
    /// ONLY idle (state 0) reads the tick.
    static func faceTreatment(state: Int, tick: Int) -> Int {
        let s = max(0, min(controlStates.count - 1, state))
        let beat = s == 0 && tick % beatPeriodTicks == 0
        return faceTreatmentTable[s * 2 + (beat ? 1 : 0)]
    }

    /// The closed region-name → face-kind table. Every `interactive: true`
    /// region in `GridLayout` MUST appear here (LINT-CONTROL-FACE + the
    /// GridLayout.selfCheck lawControlFaceTotal mirror). "brackets" ONLY
    /// where the control's content is an image (never obscure a content
    /// pixel); everything solid wears "frame".
    static let controlFaces: [String: String] = [
        // Capture / face preview (record's frame IS the CellButton ring)
        "record": "frame",
        "settingsButton": "frame",
        // Settings scene
        "setModeLive": "frame",
        "setModeFace": "frame",
        "toggleBleed": "frame",
        "toggleMirror": "frame",
        "setLibrary": "frame",
        "setArrange": "frame",
        "setResetLayout": "frame",
        "setClose": "frame",
        // The movable widget surface (spec/ui/WidgetGrid.hs WG9 — the
        // three interactive widgets). Their names are `Widget`'s raw
        // values, so the spec's vocabulary, the footprint table, the
        // lint scan and this face table all key off ONE string.
        "shutter": "frame",
        "setButton": "frame",
        "shelfButton": "frame",
        // Library (tiles show GIF content → brackets; actions solid)
        "libCell0": "brackets", "libCell1": "brackets", "libCell2": "brackets",
        "libCell3": "brackets", "libCell4": "brackets", "libCell5": "brackets",
        "libCell6": "brackets", "libCell7": "brackets", "libCell8": "brackets",
        "libShare": "frame",
        "libDelete": "frame",
        "libClose": "frame",
        // Result
        "resultShare": "frame",
        "resultKeep": "frame",
        "resultRetake": "frame",
        // Error
        "errRetry": "frame",
        "errSettings": "frame",
    ]

    /// DEBUG init-time law check (run alongside the lattice selfChecks).
    static func selfCheck() {
        assert(faceTreatmentTable.count == controlStates.count * 2, "M1 table shape")
        assert(faceTreatmentTable.allSatisfy { $0 >= 0 && $0 < faceTreatments.count },
               "M2 treatments total")
        for s in 1..<controlStates.count {
            assert(faceTreatmentTable[s * 2] == faceTreatmentTable[s * 2 + 1],
                   "M3 only idle reads the tick")
        }
        assert(goldenBeat[1] == false, "M4 tick 1 is beat-free (reduce-motion anchor)")
        for (t, b) in goldenBeat.enumerated() {
            assert(b == (t % beatPeriodTicks == 0), "M5 beat period")
        }
        assert(controlFaces.values.allSatisfy { faceKinds.contains($0) },
               "M6 faces are frame|brackets")
    }
}
